src/core/shape_inference/include/segment_max_shape_inference.hpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cmath>
+
+#include "openvino/op/segment_max.hpp"
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace v16 {
+template <class TShape, class TRShape = result_shape_t<TShape>>
+std::vector<TRShape> shape_infer(const SegmentMax* op,
+                                 const std::vector<TShape>& input_shapes,
+                                 const ITensorAccessor& tensor_accessor = make_tensor_accessor()) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 2 || input_shapes.size() == 3);
```

Comment:
```suggestion
    OPENVINO_ASSERT(input_shapes.size() == 2 || input_shapes.size() == 3);
```
Consider apply if there will be any other change in this PR